src/digest.c

```diff
@@ -139,18 +145,65 @@ void rev_memcpy(char *dst, const void *src, int len) {
 }
 
 // n.b. ripe templating to e.g. _store_from_integral<> if switching to c++
-void _store_from_int32(const uint32_t hash, char *output, const int leaveRaw) {
+SEXP _store_from_int32(const uint32_t hash, const int leaveRaw) {
+    SEXP result = R_NilValue; int output_length = sizeof(uint32_t);
+    if (leaveRaw) {
+        PROTECT(result = allocVector(RAWSXP, output_length));
+        rev_memcpy(RAW(result), &hash, output_length);
+    } else {
+        char output[output_length*2 + 1];
+        snprintf(output, output_length*2 + 1, "%08x", hash);
+        PROTECT(result = allocVector(STRSXP, 1));
+        SET_STRING_ELT(result, 0, mkChar(output));
+    }
+    UNPROTECT(1);
+    return result;
+}
+
+SEXP _store_from_int64(const uint64_t hash, const int leaveRaw) {
+    SEXP result = R_NilValue; int output_length = sizeof(uint64_t);
     if (leaveRaw) {
-        rev_memcpy(output, &hash, sizeof(uint32_t));
-    } else snprintf(output, sizeof(uint32_t)*2 + 1, "%08x", hash);
+        PROTECT(result = allocVector(RAWSXP, output_length));
+        rev_memcpy(RAW(result), &hash, output_length);
+    } else {
+        char output[output_length*2 + 1];
+        snprintf(output, output_length*2 + 1, "%016" PRIx64, hash);
+        PROTECT(result = allocVector(STRSXP, 1));
+        SET_STRING_ELT(result, 0, mkChar(output));
+    }
+    UNPROTECT(1);
+    return result;
 }
 
-void _store_from_int64(const uint64_t hash, char *output, const int leaveRaw) {
+SEXP _store_from_2xint64(const uint64_t hashlo, const uint64_t hashhi, const int leaveRaw) {
+    SEXP result = R_NilValue; int output_length = 2*sizeof(uint64_t);
     if (leaveRaw) {
-        rev_memcpy(output, &hash, sizeof(uint64_t));
-    } else snprintf(output, sizeof(uint64_t)*2 + 1, "%016" PRIx64, hash);
+        PROTECT(result = allocVector(RAWSXP, output_length));
+        rev_memcpy(RAW(result), &hashhi, 8);
+        rev_memcpy(RAW(result) + 8, &hashlo, 8);
+    } else {
+        char output[output_length*2 + 1];
+        snprintf(output, output_length*2 + 1, "%016" PRIx64 "%016" PRIx64, hashhi, hashlo);
+        PROTECT(result = allocVector(STRSXP, 1));
+        SET_STRING_ELT(result, 0, mkChar(output));
+    }
+    UNPROTECT(1);
+    return result;
 }
 
+#define FILEHASH(WHAT) if (length >= 0) { \
+    while ( ( nChar = fread( buf, 1, sizeof( buf ), fp ) ) > 0 && length > 0) { \
+        if (nChar > length) nChar = length; \
+        WHAT; \
+        length -= nChar; \
+    } \
+} else { \
+    while ( ( nChar = fread( buf, 1, sizeof( buf ), fp ) ) > 0) { \
+        WHAT; \
+    } \
+} \
+fclose(fp);
+
```

Comment:
Ewww.

Reply:
yeeeah, agree that particular dangler is a few stray keystrokes away from a mess.